app/services/check_assignment_status.rb

```diff
@@ -1,88 +1,126 @@
+
+        pagenames.each_slice(50) do |batch|
+          api = WikiApi.new(wiki)
+          page_infos = api.get_page_info(batch)
 
-  def update_status
-    @sandboxes.each do |sandbox_key, page_details|
-      info = WikiApi.new(page_details[:wiki]).get_page_info page_details[:pagename]
-      new_status = page_status(info)
-      @assignment.update_sandbox_status(sandbox_key, new_status)
+          process_batch(entries, batch, page_infos)
+        end
+      end
     end
-    # Do any already have corresponding Article records?
-    # If not, have any been created?
-    # If so, what namespaces are they in? what are their current titles? how big are they?
-  end
 
-  # Takes a hash of page info from the MediaWiki API and returns a status for
-  # the Assignment record
-  def page_status(page_info)
-    return AssignmentPipeline::SandboxStatuses::DOES_NOT_EXIST unless page_present?(page_info)
-
-    case page_info['pages'].values.first['ns']
-    when Article::Namespaces::USER
-      AssignmentPipeline::SandboxStatuses::EXISTS_IN_USERSPACE
-    when Article::Namespaces::DRAFT
-      AssignmentPipeline::SandboxStatuses::EXISTS_IN_DRAFT_SPACE
-    when Article::Namespaces::MAINSPACE
-      AssignmentPipeline::SandboxStatuses::EXISTS_IN_MAINSPACE
-    else
-      AssignmentPipeline::SandboxStatuses::EXISTS_ELSEWHERE
+    def process_batch(entries, batch, page_infos)
+      return unless page_infos && page_infos['pages']
+
+      pages_by_normalized_title = build_pages_by_normalized_title(page_infos)
+
+      Assignment.transaction do
+        update_assignments_for_batch(entries, batch, pages_by_normalized_title)
+      end
     end
-  end
 
-  def page_present?(page_info)
-    return false unless page_info.present?
-    page_info.dig('pages', '-1', 'missing').nil?
+    def build_pages_by_normalized_title(page_infos)
+      page_infos['pages'].each_with_object({}) do |(_, page), hash|
+        normalized_title = page['title'].tr(' ', '_')
+        page['present'] = !page.key?('missing')
+        hash[normalized_title] = page
+      end
+    end
+
+    def update_assignments_for_batch(entries, batch, pages_by_normalized_title)
+      batch.each do |pagename|
+        normalized_pagename = pagename.tr(' ', '_')
+        page_data = pages_by_normalized_title[normalized_pagename]
+        status = determine_status(page_data)
+
+        update_entries_for_pagename(entries, pagename, status)
+      end
+    end
+
+    def determine_status(page_data)
+      if page_data
+        status_from_namespace(page_data['ns'], page_data['present'])
+      else
+        AssignmentPipeline::SandboxStatuses::DOES_NOT_EXIST
+      end
+    end
+
+    def update_entries_for_pagename(entries, pagename, status)
+      entries.select { |e| e[:pagename] == pagename }.each do |entry|
+        entry[:assignment].update_sandbox_status(entry[:key], status)
+      end
+    end
+
+    def find_page_data(page_infos, pagename)
```

Comment:
Comments showing what the page_infos data looks like would be useful here, since the code is doing nontrivial things to search through a specific expected data structure.